File: server-manager/src/main/java/com/mcmanager/server/web/controller/BackupController.java
package com.mcmanager.server.web.controller;

import com.mcmanager.core.model.BackupEntry;
import com.mcmanager.core.model.InstanceConfig;
import com.mcmanager.server.instance.ServerInstanceManager;
import com.mcmanager.server.service.BackupService;
import io.javalin.http.Context;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.List;
import java.util.Map;

/**
 * REST endpoints for instance backups: list the audit trail, trigger a manual
 * backup, restore an archive over the instance directory, and configure how
 * many backups are kept.
 */
public class BackupController {

    private final BackupService backupService;
    private final ServerInstanceManager instanceManager;

    public BackupController(BackupService backupService, ServerInstanceManager instanceManager) {
        this.backupService = backupService;
        this.instanceManager = instanceManager;
    }

    /** GET /api/instances/{id}/backups */
    public void listBackups(Context ctx) {
        String instanceId = ctx.pathParam("id");
        List<BackupEntry> backups = backupService.listBackups(instanceId);
        ctx.json(Map.of("backups", backups));
    }

    /** POST /api/instances/{id}/backups — creates a manual backup */
    public void createBackup(Context ctx) {
        String instanceId = ctx.pathParam("id");
        try {
            BackupEntry entry = backupService.createBackup(instanceId, BackupEntry.TRIGGER_MANUAL);
            ctx.status(201).json(entry);
        } catch (IllegalArgumentException e) {
            ctx.status(404).result(e.getMessage());
        } catch (IOException e) {
            ctx.status(500).result("Backup creation failed: " + e.getMessage());
        }
    }

    /** POST /api/instances/{id}/backups/{backupId}/restore */
    public void restoreBackup(Context ctx) {
        String instanceId = ctx.pathParam("id");
        String backupId = ctx.pathParam("backupId");
        try {
            backupService.restoreBackup(instanceId, backupId);
            ctx.json(Map.of("ok", true, "message", "Backup restored successfully."));
        } catch (FileNotFoundException | IllegalArgumentException e) {
            ctx.status(404).result(e.getMessage());
        } catch (IOException e) {
            ctx.status(500).result("Restore failed: " + e.getMessage());
        }
    }

    /** POST /api/instances/{id}/backups/retention — body: {retention: N} */
    public void setRetention(Context ctx) {
        String instanceId = ctx.pathParam("id");
        RetentionRequest body;
        try {
            body = ctx.bodyAsClass(RetentionRequest.class);
        } catch (RuntimeException e) {
            ctx.status(400).result("Invalid JSON body");
            return;
        }
        if (body == null || body.retention == null) {
            ctx.status(400).result("retention is required");
            return;
        }
        try {
            instanceManager.getInstance(instanceId); // 404 for unknown ids
            if (body.retention < InstanceConfig.MIN_BACKUP_RETENTION
                    || body.retention > InstanceConfig.MAX_BACKUP_RETENTION) {
                ctx.status(400).result("retention must be between "
                        + InstanceConfig.MIN_BACKUP_RETENTION + " and "
                        + InstanceConfig.MAX_BACKUP_RETENTION);
                return;
            }
            int deleted = backupService.setRetention(instanceId, body.retention);
            ctx.json(Map.of("retention", body.retention, "deletedBackups", deleted));
        } catch (IllegalArgumentException e) {
            ctx.status(404).result(e.getMessage());
        } catch (IOException e) {
            ctx.status(500).result("Failed to save retention: " + e.getMessage());
        }
    }

    public static class RetentionRequest {
        public Integer retention;
    }
}

File: crates/zircon-server/src/web/controllers/backup_controller.rs
//! REST endpoints for instance backups: list the audit trail, trigger a manual
//! backup, restore an archive over the instance directory, and configure how
//! many backups are kept.
//!
//! Port of `com.mcmanager.server.web.controller.BackupController`.

use axum::extract::{Path, State};
use axum::Json;

use serde::Deserialize;

use crate::web::app::{ApiError, AppState};
use crate::web::views;

/// GET /api/instances/{id}/backups
pub async fn list_backups(
    State(state): State<AppState>,
    Path(instance_id): Path<String>,
) -> Json<serde_json::Value> {
    let backups = state.backup.list_backups(&instance_id);
    let mapped: Vec<serde_json::Value> = backups.iter().map(views::backup_entry_to_map).collect();
    Json(serde_json::json!({ "backups": mapped }))
}

/// POST /api/instances/{id}/backups — creates a manual backup.
pub async fn create_backup(
    State(state): State<AppState>,
    Path(instance_id): Path<String>,
) -> Result<(axum::http::StatusCode, Json<serde_json::Value>), ApiError> {
    let entry = state.backup.create_backup(&instance_id, "manual").await?;
    Ok((
        axum::http::StatusCode::CREATED,
        Json(views::backup_entry_to_map(&entry)),
    ))
}

/// POST /api/instances/{id}/backups/{backup_id}/restore
pub async fn restore_backup(
    State(state): State<AppState>,
    Path((instance_id, backup_id)): Path<(String, String)>,
) -> Result<Json<serde_json::Value>, ApiError> {
    state
        .backup
        .restore_backup(&instance_id, &backup_id)
        .await?;
    Ok(Json(serde_json::json!({
        "ok": true,
        "message": "Backup restored successfully."
    })))
}

/// POST /api/instances/{id}/backups/retention — body: {retention: N}
pub async fn set_retention(
    State(state): State<AppState>,
    Path(instance_id): Path<String>,
    Json(body): Json<RetentionRequest>,
) -> Result<Json<serde_json::Value>, ApiError> {
    let Some(retention) = body.retention else {
        return Err(ApiError::BadRequest("retention is required".to_string()));
    };
    if !(1..=100).contains(&retention) {
        return Err(ApiError::BadRequest(
            "retention must be between 1 and 100".to_string(),
        ));
    }
    let deleted = state.backup.set_retention(&instance_id, retention)?;
    Ok(Json(
        serde_json::json!({ "retention": retention, "deletedBackups": deleted }),
    ))
}

#[derive(Debug, Deserialize)]
pub struct RetentionRequest {
    pub retention: Option<i32>,
}
